### ISSUE TITLE
WALs retired unexpectedly [2.32]

### ISSUE THREAD

**Issue by psuderevsky** [pgbackrest/pgbackrest#1402]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version:
2.28 - backups done and WALs archived
2.32 - issue spotted

2. PostgreSQL version:
11.11

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
CentOS Linux release 7.9.2009 (3.10.0-1160.21.1.el7.x86_64)

4. Did you install pgBackRest from source or from a package?
package

5. Please attach the following as applicable:
pgbackrest.conf
```
[someplatform]
pg1-path=/postgres/11/pgdata
repo1-cipher-pass=somepass
repo1-cipher-type=aes-256-cbc
repo1-s3-bucket=bucketname
repo1-s3-endpoint=s3.us-east-2.amazonaws.com
repo1-path=/pgbackrest
repo1-s3-region=us-east-2
repo1-s3-key=somekey
repo1-s3-key-secret=somesecret
repo1-type=s3
repo1-s3-verify-tls=n
repo1-retention-full-type=time
repo1-retention-full=30
repo1-retention-diff=4
process-max=4
start-fast=n
stop-auto=n
backup-standby=n
compress=y
compress-level=5
spool-path=/dbwal/backrest
buffer-size=8388608

[someplatform:backup]
pg1-path=/postgres/11/pgdata
pg2-host=localhost
pg2-host-port=2293
pg2-path=/postgres/11/pgdata
compress-level=6
process-max=32

[someplatform:restore]
process-max=64 

[someplatform:archive-get]
archive-async=y
archive-get-queue-max=2GB
process-max=6

[someplatform:archive-push]
archive-async=y
archive-push-queue-max=200GB
process-max=6
```
archive command
```
archive_command: pgbackrest --stanza=someplatform archive-push %p
```

7. Describe the issue:
Here are some oldest pgbackrest backups:
```
stanza: someplatform
    status: ok
    cipher: aes-256-cbc

    db (current)
        wal archive min/max (11): 00000056000019FF00000051/00000069000018B700000056

        full backup: 20210410-190004F
            timestamp start/stop: 2021-04-10 19:00:04 / 2021-04-10 19:34:07
            wal start/stop: 00000056000018E300000049 / 00000056000018E30000004C
            database size: 1586.7GB, database backup size: 1586.7GB
            repo1: backup set size: 192.4GB, backup size: 192.4GB

        full backup: 20210417-190004F
            timestamp start/stop: 2021-04-17 19:00:04 / 2021-04-17 19:29:53
            wal start/stop: 00000056000019BF000000CC / 00000056000019BF000000CE
            database size: 1324.3GB, database backup size: 1324.3GB
            repo1: backup set size: 183.4GB, backup size: 183.4GB

        full backup: 20210424-190005F
            timestamp start/stop: 2021-04-24 19:00:05 / 2021-04-24 19:31:10
            wal start/stop: 00000056000019F600000063 / 00000056000019F600000066
            database size: 1435.2GB, database backup size: 1435.2GB
            repo1: backup set size: 194.6GB, backup size: 194.6GB

        full backup: 20210501-190004F
            timestamp start/stop: 2021-05-01 19:00:04 / 2021-05-01 19:34:27
            wal start/stop: 00000056000019FD000000BA / 00000056000019FD000000BD
            database size: 1445.7GB, database backup size: 1445.7GB
            repo1: backup set size: 195.8GB, backup size: 195.8GB

        full backup: 20210508-190005F
            timestamp start/stop: 2021-05-08 19:00:05 / 2021-05-08 19:31:46
            wal start/stop: 0000005700001A0400000047 / 0000005700001A040000004A
            database size: 1446.2GB, database backup size: 1446.2GB
            repo1: backup set size: 195.6GB, backup size: 195.6GB

        diff backup: 20210508-190005F_20210509-190005D
            timestamp start/stop: 2021-05-09 19:00:05 / 2021-05-09 19:05:12
            wal start/stop: 0000005700001A04000000D9 / 0000005700001A04000000D9
            database size: 1446.3GB, database backup size: 14GB
            repo1: backup set size: 195.7GB, backup size: 1.4GB
            backup reference list: 20210508-190005F

        incr backup: 20210508-190005F_20210510-063004I
            timestamp start/stop: 2021-05-10 06:30:04 / 2021-05-10 06:33:38
            wal start/stop: 0000005700001A050000001F / 0000005700001A050000001F
            database size: 1446.4GB, database backup size: 17.9GB
            repo1: backup set size: 195.7GB, backup size: 1.8GB
            backup reference list: 20210508-190005F, 20210508-190005F_20210509-190005D
```
Oldest WAL is **00000056000019FF00000051** while backups prior to **20210508-190005F** require older WALs.

According log, WAL required for the oldest backup was pushed on April 10:
```
-------------------PROCESS START-------------------
2021-04-10 19:10:04.697 P00   INFO: archive-push:async command begin 2.32: [/postgres/11/pgdata/pg_wal] --archive-async --archive-push-queue-max=214748364800 --buffer-size=8388608 --compress-level=5 --exec-id=144432-4bda2370 --log-level-console=off --log-level-stderr=off --pg1-path=/postgres/11/pgdata --process-max=6 --repo1-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo1-path=/pgbackrest --repo1-s3-bucket=bucketname --repo1-s3-endpoint=s3.us-east-2.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-east-2 --no-repo1-s3-verify-tls --repo1-type=s3 --spool-path=/dbwal/backrest --stanza=someplatform
2021-04-10 19:10:04.698 P00   INFO: push 1 WAL file(s) to archive: 00000056000018E300000049
2021-04-10 19:10:06.218 P00   INFO: archive-push:async command end: completed successfully (1522ms)

-------------------PROCESS START-------------------
2021-04-10 19:20:05.022 P00   INFO: archive-push:async command begin 2.32: [/postgres/11/pgdata/pg_wal] --archive-async --archive-push-queue-max=214748364800 --buffer-size=8388608 --compress-level=5 --exec-id=151142-2a5b6cf2 --log-level-console=off --log-level-stderr=off --pg1-path=/postgres/11/pgdata --process-max=6 --repo1-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo1-path=/pgbackrest --repo1-s3-bucket=bucketname --repo1-s3-endpoint=s3.us-east-2.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-east-2 --no-repo1-s3-verify-tls --repo1-type=s3 --spool-path=/dbwal/backrest --stanza=someplatform
2021-04-10 19:20:05.023 P00   INFO: push 1 WAL file(s) to archive: 00000056000018E30000004A
2021-04-10 19:20:06.198 P00   INFO: archive-push:async command end: completed successfully (1177ms)

-------------------PROCESS START-------------------
2021-04-10 19:30:05.098 P00   INFO: archive-push:async command begin 2.32: [/postgres/11/pgdata/pg_wal] --archive-async --archive-push-queue-max=214748364800 --buffer-size=8388608 --compress-level=5 --exec-id=158584-24e670c9 --log-level-console=off --log-level-stderr=off --pg1-path=/postgres/11/pgdata --process-max=6 --repo1-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo1-path=/pgbackrest --repo1-s3-bucket=bucketname --repo1-s3-endpoint=s3.us-east-2.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-east-2 --no-repo1-s3-verify-tls --repo1-type=s3 --spool-path=/dbwal/backrest --stanza=someplatform
2021-04-10 19:30:05.099 P00   INFO: push 1 WAL file(s) to archive: 00000056000018E30000004B
2021-04-10 19:30:06.361 P00   INFO: archive-push:async command end: completed successfully (1264ms)

-------------------PROCESS START-------------------
2021-04-10 19:34:07.902 P00   INFO: archive-push:async command begin 2.32: [/postgres/11/pgdata/pg_wal] --archive-async --archive-push-queue-max=214748364800 --buffer-size=8388608 --compress-level=5 --exec-id=161373-3e8bf098 --log-level-console=off --log-level-stderr=off --pg1-path=/postgres/11/pgdata --process-max=6 --repo1-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo1-path=/pgbackrest --repo1-s3-bucket=bucketname --repo1-s3-endpoint=s3.us-east-2.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-east-2 --no-repo1-s3-verify-tls --repo1-type=s3 --spool-path=/dbwal/backrest --stanza=someplatform
2021-04-10 19:34:07.903 P00   INFO: push 2 WAL file(s) to archive: 00000056000018E300000049.00000028.backup...00000056000018E30000004C
2021-04-10 19:34:08.722 P00   INFO: archive-push:async command end: completed successfully (821ms)
```

Oldest existing in archive WAL was pushed on May 4.
```
-------------------PROCESS START-------------------
2021-05-04 06:45:19.077 P00   INFO: archive-push:async command begin 2.32: [/postgres/11/pgdata/pg_wal] --archive-async --archive-push-queue-max=214748364800 --buffer-size=8388608 --compress-level=5 --exec-id=46899-93995cab --log-level-console=off --log-level-stderr=off --pg1-path=/postgres/11/pgdata --process-max=6 --repo1-cipher-pass=<redacted> --repo1-cipher-type=aes-256-cbc --repo1-path=/pgbackrest --repo1-s3-bucket=bucketname --repo1-s3-endpoint=s3.us-east-2.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-ke
y-secret=<redacted> --repo1-s3-region=us-east-2 --no-repo1-s3-verify-tls --repo1-type=s3 --spool-path=/dbwal/backrest --stanza=someplatform
2021-05-04 06:45:19.081 P00   INFO: push 1 WAL file(s) to archive: 00000056000019FF00000051
2021-05-04 06:45:20.004 P00   INFO: archive-push:async command end: completed successfully (928ms)
```

Since no explicit WAL retention was configured, there is a suspicion that WALs were unexpectedly retired during deprecated incr/diff backups retentions. This is one of the incremental backups that was retentioned (**00000056000019FF00000051** is the next after this backup stop archive WAL **00000056000019FF00000050**):
```
2021-05-04 06:35:19.668 P00   INFO: execute non-exclusive pg_stop_backup() and wait for all WAL segments to archive
2021-05-04 06:35:19.870 P00   INFO: backup stop archive = 00000056000019FF00000050, lsn = 19FF/502C10F0
2021-05-04 06:35:19.960 P00   INFO: check archive for segment(s) 00000056000019FF00000050:00000056000019FF00000050
2021-05-04 06:35:20.602 P00   INFO: new backup label = 20210501-190004F_20210504-063004I
2021-05-04 06:35:21.752 P00   INFO: backup command end: completed successfully (319716ms)
```
Here is a corresponding expire event:
```
-------------------PROCESS START-------------------
2021-05-12 06:35:47.650 P00   INFO: expire command begin 2.32: --buffer-size=8388608 --exec-id=13626-472b793d --log-level-console=info --repo1-cipher-pass=<redacted> --repo1-cipher-type
=aes-256-cbc --repo1-path=/pgbackrest --repo1-retention-diff=4 --repo1-retention-full=30 --repo1-retention-full-type=time --repo1-s3-bucket=bucketname --repo
1-s3-endpoint=s3.us-east-2.amazonaws.com --repo1-s3-key=<redacted> --repo1-s3-key-secret=<redacted> --repo1-s3-region=us-east-2 --no-repo1-s3-verify-tls --repo1-type=s3 --stanza=someplatform
2021-05-12 06:35:48.956 P00   INFO: expire diff backup set repo1: 20210501-190004F_20210502-190005D, 20210501-190004F_20210503-063004I, 20210501-190004F_20210503-120005I, 20210501-190004F_20210503-150005I, 20210501-190004F_20210504-063004I, 20210501-190004F_20210504-120005I, 20210501-190004F_20210504-150005I
2021-05-12 06:35:49.166 P00   INFO: remove expired backup repo1: 20210501-190004F_20210504-150005I
2021-05-12 06:35:50.414 P00   INFO: remove expired backup repo1: 20210501-190004F_20210504-120005I
2021-05-12 06:35:51.673 P00   INFO: remove expired backup repo1: 20210501-190004F_20210504-063004I
2021-05-12 06:35:53.210 P00   INFO: remove expired backup repo1: 20210501-190004F_20210503-150005I
2021-05-12 06:35:54.230 P00   INFO: remove expired backup repo1: 20210501-190004F_20210503-120005I
2021-05-12 06:35:55.791 P00   INFO: remove expired backup repo1: 20210501-190004F_20210503-063004I
2021-05-12 06:35:57.182 P00   INFO: remove expired backup repo1: 20210501-190004F_20210502-190005D
2021-05-12 06:35:58.448 P00   INFO: time-based archive retention not met for repo1 - archive logs will not be expired
2021-05-12 06:35:58.448 P00   INFO: expire command end: completed successfully (10798ms)
```

**Comment by dwsteele** [comment]
> time-based archive retention not met for repo1 - archive logs will not be expired

This message is output when WAL expiration is skipped, so I don't think it happened in that execution at least.

I'm wondering if this is a reporting issue. Have you checked the repo directly?
```
pgbackrest repo-ls --recurse --sort=asc archive/[stanza] | head
```

**Comment by psuderevsky** [comment]
Sure, I've checked and there are no WALs prior to **00000056000019FF00000051**.
```
~$ pgbackrest repo-ls --recurse --sort=asc archive/someplatform| head -4
11-1
11-1/00000056000019FF
11-1/00000056000019FF/00000056000019FF00000051-468df1bb9f091e77dcdfb709c0fe693a184218b4.gz
11-1/00000056000019FF/00000056000019FF00000052-5db18ba8069b8f00063fc13a0a079ac637a60756.gz
```

> This message is output when WAL expiration is skipped, so I don't think it happened in that execution at least.

Anyway, WAL expiration shouldn't retire WALs required for existing backups restoring. So if they were removed automatically by expire job, then only the backups removal (with associated WALs) could cause the WALs to be deleted unexpectedly.

**Comment by dwsteele** [comment]
Also, if you can provide the full expire log we'll have a look.

**Comment by psuderevsky** [comment]
[someplatform-expire.log](https://github.com/pgbackrest/pgbackrest/files/6504213/someplatform-expire.log)
Log attached.

> Any chance these files were removed by a bucket policy or accident?

I believe no.

There is an assumption that this could be caused by presence of WAL for the timelines higher than current one in the archive, but rollback to the old timeline (rollback from another repository, backups of this timeline in current repo had been removed already) happened much earlier than this issue occurred. More precisely:
1. repo 1 containing WALs for TLs 101-105, cluster is running on TL 105
2. restore cluster from repo 2 to a backup with TL 85, promote to TL 86 and making new backup full in repo 1
Yeah, I had to remove backups from repo 1 after restore or promote it to TL > 105 before starting WALs archiving, but still - not clear how that could cause the issue.

**Comment by dwsteele** [comment]
There is nothing too interesting in the log. It would need to be at detail level to be helpful. We are considering raising the level of these messages by default, but that would not happen until 2.34.

As for the rest, well that's a lot to digest. You talk about multiple repos, which are not supported in < 2.33. You are removing backups for the repo, but there are no adhoc expires in the logs, so I guess this was done manually.

Basically, I'm not sure how we can reasonably reproduce anything that is going on here. Increase your log level to at least `detail` so we can have a look if this happens again.

**Comment by psuderevsky** [comment]
Hasn't happened since that time. I don't see any way to investigate this anymore, so I'll reopen the case if something similar happens again.
